### PR TITLE
[SPARK-36754][SQL] ArrayIntersect handle duplicated Double.NaN and Float.NaN

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
For query
```
select array_intersect(array(cast('nan' as double), 1d), array(cast('nan' as double)))
```
This returns [NaN], but it should return [].
This issue is caused by `OpenHashSet` can't handle `Double.NaN` and `Float.NaN` too.
In this pr fix this based on https://github.com/apache/spark/pull/33955

### Why are the changes needed?
Fix bug

### Does this PR introduce _any_ user-facing change?
ArrayIntersect won't show equal `NaN` value

### How was this patch tested?
Added UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -3826,32 +3826,43 @@ case class ArrayIntersect(left: Expression, right: Expression) extends ArrayBina
...
+                hsResult.addNull()
               }
             } else {
               val elem = array1.get(i, elementType)
-              if (hs.contains(elem) && !hsResult.contains(elem)) {
-                arrayBuffer += elem
-                hsResult.add(elem)
+              if (isNaN(elem)) {
+                if (hs.containsNaN() && !hsResult.containsNaN()) {
+                  arrayBuffer += elem
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3858]: For this, let's wait a little bit for the decision at the first PR.
- https://github.com/apache/spark/pull/33955/files#r708570515

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2/3.1/3.0!

**Comment by karenfeng** [comment]
For clarification @AngersZhuuuu: the PR description says:

> For query
> ```
> select array_intersect(array(cast('nan' as double), 1d), array(cast('nan' as double)))
> ```
> This returns [NaN], but it should return [].

Is this the right way around? It seems like we now correctly return `[NaN]`, but previously incorrectly returned `[]`.

**Comment by AngersZhuuuu** [comment]
> For clarification @AngersZhuuuu: the PR description says:
> 
> > For query
> > ```
> > select array_intersect(array(cast('nan' as double), 1d), array(cast('nan' as double)))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This returns [NaN], but it should return [].
> 
> Is this the right way around? It seems like we now correctly return `[NaN]`, but previously incorrectly returned `[]`.

Oh， sorry for the mistake. Correct is we should return [NaN]

**Pull request closed without merging**